lmdeploy/serve/turbomind/deploy.py

```diff
@@ -379,38 +130,351 @@ def permute(x: torch.Tensor):
+                    elif i == 0:
+                        param = get_param(param_name, param_data.size())
+                        param.data = param_data
+                del ckpt
+
+            for name, param in model_params.items():
+                # transpose all weights as TurboMind is expecting column-major
+                # (output_dims, input_dims) -> (input_dims, output_dims)
+                key = name.split('.')[-2]
+                if key in ['w1', 'w3', 'wq', 'wk', 'wv', 'w2', 'wo']:
+                    param.data = param.data.t()
+                    if key in ['wq', 'wk']:
+                        param.data = reverse_permute(param.data)
+
+            params = {}
+            # prefill transformer_layer
+            transformer_layer = []
+            for i in range(self.num_layer):
+                # attn
+                _attn_param = []
+                for suffix in ['weight', 'bias']:
+                    for key in ['wq', 'wk', 'wv', 'wo']:
+                        tensor = model_params.pop(
+                            f'layers.{i}.attention.{key}.{suffix}', None)
+                        tensor = tensor.t() if tensor is not None else None
+                        _attn_param.append(tensor)
+                attn = AttentionLayer(*_attn_param)
+                # ffn
+                _ffn_param = []
+                for key in ['w1', 'w2', 'w3']:
+                    tensor = model_params.pop(
+                        f'layers.{i}.feed_forward.{key}.weight')
+                    _ffn_param.append(tensor.t())
+                ffn = FeedForwardLayer(*_ffn_param)
+                # norm
+                attn_norm = model_params.pop(
+                    f'layers.{i}.attention_norm.weight')
+                ffn_norm = model_params.pop(f'layers.{i}.ffn_norm.weight')
+                transformer_layer.append(
+                    TransformerLayer(attn, ffn, attn_norm, ffn_norm))
+            params['layers'] = transformer_layer
+
+            # other weights
+            others = ['tok_embeddings.weight', 'norm.weight', 'output.weight']
+            for key in others:
+                params[key] = model_params.pop(key)
+            self.params = params
+
+        def layers(self):
+            for layer_id in range(self.num_layer):
+                yield self.params['layers'][layer_id]
+
+    def bins(self):
+        for bin in self.bin_files:
+            yield bin
+
+
+class HfModel:
+    """Model reader to read model in hf format.
+
+    Args:
+        model_path (str): the path of the directory where the model weight
+          files are
+        tokenizer_path (str): the path of the tokenizer model
+    """
+
+    def __init__(self, model_path: str, tokenizer_path: str, **kwargs):
+        self.model_path = model_path
+        self.tokenizer_path = tokenizer_path
+        if 'bin_files' in kwargs:
+            bin_files = kwargs['bin_files']
+        else:
+            _suffixes = ['.safetensors', '.bin']
```

Comment:
'.safetensors' and '.bin' are the same weight files but in different formats. We cannot put them all in 'bin_files' list